### ISSUE TITLE
Crater GFX in 2nd Level is wrong / blue enemy GFX glitch

### ISSUE THREAD

**Issue by PhantombrainM** [MiSTer-devel/Arcade-MoonPatrol_MiSTer#15]
The craters in 2nd Level at the bottom show the background hills while it should just be a solid gfx.
Also the blue triangle enemy which can shoot crateers has a gfx glitch in animation.

I noticed Mame needs a new ROM file for mpatrol which maybe should be integrated into the .mra / core as well?
it's mp_7621-5.7h (512 bytes)

**Comment by Real-Dampfwalze** [comment]
Can confirm this problem.

**Comment by febs** [comment]
Please test https://github.com/febs/Arcade-MoonPatrol_MiSTer/raw/refs/heads/test/test_releases/Arcade-MoonPatrol-test-fix-issue15.rbf and report if that solves the issue. Thanks,

**Comment by Real-Dampfwalze** [comment]
Tested it and it fixes the background problem.
But there are now graphical glitches (lines missing in the sprites, lines appearing above or under the sprites) with the Moon Buggy, its tires when it explodes, mines on the ground and the little car in round 2 that tries to ram you from behind.
Rocks, turrets and volcanoes are OK though.
Thank you for your work. 👍

**Comment by febs** [comment]
You're more than welcome, although I am just being an AI "meat proxy" here. I'm guiding the AI but the real work is done by that. Quite impressively admittedly.

I fed it your feedback and after a couple minutes investigating it responded with

> I checked the changes: the sprite renderer itself was not modified. The experimental patch changed both:
> 
>   - how the solid lower portions of backgrounds are generated;
>   - the priority between the three background layers.
> 
>   The Moon Buggy is assembled from four separate hardware sprites, so a one-scanline alignment problem can produce exactly the reported gaps and stray lines. Moon Patrol hardware
>   documentation confirms this multi-sprite construction.
> 
>   The safest next step is to make two diagnostic RBFs:
> 
>   1. solid-background fix only, retaining the original layer priority;
>   2. layer-priority change only, retaining the original background renderer.
> 
>   That will identify which half causes the regression instead of guessing. A photograph or short video from the tester would also help establish whether the lines are erroneous background
>   pixels or displaced sprite rows.
>

Would you mind providing such a short video to help Codex move forward on the correct debugging path? Thanks for your help.

**Comment by HerrBerzerk** [comment]
Well, some of these sprite glitches was in the core before the "solid background" changes, just not so many and so obvious. What was before, was, that sometimes near the falling bombs were sprite glitches, but the buggy, beside from the explosion, was OK. The Background problems with craters and letters was, because the green background was not there/not hight enough.

In my oppionion, the stones are also not 100%, there is always a transparent pixels in them, like in the mines also.

Maybe this helps, bad quality because of CRT and somehow github does not allow big files... maybe Dampwalze can help out....

https://github.com/user-attachments/assets/cba67115-d0ff-4d0b-b689-632791ccfdbc

**Comment by Real-Dampfwalze** [comment]
Everything HerrBerzerk said and I noticed that there are sometimes also glitches in the flying enemies.

I made a first round full playthrough up to the second round till the volcanos begin to appear.
You may compare the video to MAME longplays on YouTube to spot the differences.

[https://youtu.be/YfrBWkAERYY](https://youtu.be/YfrBWkAERYY)

**Comment by febs** [comment]
> Everything HerrBerzerk said and I noticed that there are sometimes also glitches in the flying enemies.
> 
> I made a first round full playthrough up to the second round till the volcanos begin to appear. You may compare the video to MAME longplays on YouTube to spot the differences.
> 
> https://youtu.be/YfrBWkAERYY

Would you mind please providing timestamps where the issues specifically and timely happen? 
Specifically when any of these happen and which of them.

* moon buggy/front wheel corruption
* missing or extra line of flying objects  
* a false bomb/line fragment
* mine or rear-car corruption

 I could find an evident example of #2 at 0.31, but it would help to get as much detailed infos as possible.

Thank you very much - and by the way, congratulations for the bloody high score! I still have to go past 13k LOL.

**Comment by Real-Dampfwalze** [comment]
OK, I try my very best.

So the Moon Buggy is corrupted the whole time. In the missing line pixels wander and often even under the Buggy.

Normal Buggy in Mame

<img width="174" height="120" alt="Image" src="https://github.com/user-attachments/assets/8cea2384-fccd-4046-8fad-d32f0957cadc" />



Buggy in Mister with lines also under Buggy Timestamp 00:38

<img width="270" height="222" alt="Image" src="https://github.com/user-attachments/assets/ec13c066-bf1c-4a38-9d99-1af9809f0833" />



Missing or extra line in flying object Timestamp 00:30

<img width="124" height="102" alt="Image" src="https://github.com/user-attachments/assets/9db23078-0999-40c3-b3fa-bbb3b9994e96" />



Wrong lines in explosions from flying bombs Timestamp 01:12

<img width="306" height="219" alt="Image" src="https://github.com/user-attachments/assets/e531688b-9cff-49c1-838b-8974aa8cf811" />



Wrong lines in crater explosions from flying bombs Timestamp 01:39

<img width="316" height="246" alt="Image" src="https://github.com/user-attachments/assets/c5e45b05-160c-4f0a-af13-9da8832ce0f8" />



Rear Car is missing a line but has a line left of it all the time it is on screen.
Rear Car normal in Mame

<img width="211" height="126" alt="Image" src="https://github.com/user-attachments/assets/f9c5b316-011a-487f-a62a-c9b6c898eab2" />



Rear Car with line corruption Timestamp 08:06-08:14

<img width="186" height="140" alt="Image" src="https://github.com/user-attachments/assets/7f699aeb-dadc-46e2-bbd9-8b775c1cffe0" />



The Rocks seem OK to me, but the color for the shading is wrong.
It is near black on Mister and should be just a darker brown.

Rocks in Mame

<img width="146" height="115" alt="Image" src="https://github.com/user-attachments/assets/bb860db3-1003-4c69-8403-bc1f438c2169" />



Rocks in Mister

<img width="311" height="222" alt="Image" src="https://github.com/user-attachments/assets/4c5f2e60-a573-4c34-9008-1e927fe7009a" />

**Comment by Real-Dampfwalze** [comment]
PS The Rock color issue could be none, because I'm not sure if colors in MAME are OK. They seem to be a bit pale. 
PPS I checked with the Hamster Arcade Archive version and their rocks look to be between MAME and Mister brightness. So they maybe OK in Mister.

**Comment by HerrBerzerk** [comment]
Beside that there are errors in the mines, extra dots, missing lines or extre lines at 2:59, 3.04 and 3.07!

<img width="894" height="488" alt="Image" src="https://github.com/user-attachments/assets/62390307-6597-4743-8b02-c6e041be54b3" />

<img width="1021" height="577" alt="Image" src="https://github.com/user-attachments/assets/6d02f99d-6dee-43af-82b7-9e508334e553" />

<img width="1290" height="713" alt="Image" src="https://github.com/user-attachments/assets/d88bab21-f5e3-42c1-b45a-7dfcbead624f" />

**Comment by febs** [comment]
Thank you everyone for the very deep insights!

The new build changes the renderer, so every sprite row is calculated independently for each scanline, preventing moving or reused sprite slots from retaining an incorrect row phase. It also resets the sprite renderer deterministically.

You can find the test release here. https://github.com/febs/Arcade-MoonPatrol_MiSTer/raw/refs/heads/test/test_releases/Arcade-MoonPatrol-test-sprite-row-timing.rbf

Thanks!

**Comment by Real-Dampfwalze** [comment]
I tested the new built and I can't spot anymore sprite corruption. To me it seems perfect now! 👍🙂🙂🙂
Thank you very much! 
I don't know if your work via AI will conflict with the cycle perfect FPGA idea, but I think you should ask if you build could be put into the official repository for download. 👍

If you feel like it, there is another "old" core that needs a fix: Frogger. The default music that plays (after the jingles are over you hear when you placed a frog in his home) plays to fast after the first frog you safed. This bugs me since 3 years or so and it has never been fixed. 🙁

**Comment by HerrBerzerk** [comment]
Absolutely... this has to go to update_all and it is the same for me with Frogger... I love it but it is unplayable on the Mister :-(

**Comment by febs** [comment]
I submitted the changes to the main repository. Thank you so much!
Will eventually pick up frogger one day. Thanks,
